pkg/parse/parse.go

```diff
@@ -589,11 +590,30 @@ func SystemContextFromOptions(c *cobra.Command) (*types.SystemContext, error) {
 		ctx.RegistriesDirPath = regConfDir
 	}
 	ctx.DockerRegistryUserAgent = fmt.Sprintf("Buildah/%s", buildah.Version)
-	if os, err := c.Flags().GetString("override-os"); err == nil {
+
+	if os, err := c.Flags().GetString("override-os"); err == nil && os != runtime.GOOS {
+		ctx.OSChoice = os
+	} else if os, err := c.Flags().GetString("os"); err == nil && os != runtime.GOOS {
 		ctx.OSChoice = os
+	} else {
+		ctx.OSChoice = runtime.GOOS
+	}
+
+	if arch, err := c.Flags().GetString("override-arch"); err == nil && arch != runtime.GOARCH {
+		ctx.ArchitectureChoice = arch
+	} else if arch, err := c.Flags().GetString("arch"); err == nil && arch != runtime.GOARCH {
+		ctx.ArchitectureChoice = arch
+	} else {
+		ctx.ArchitectureChoice = runtime.GOARCH
 	}
-	if arch, err := c.Flags().GetString("override-arch"); err == nil {
+
+	if pf, err := c.Flags().GetString("platform"); err == nil && pf != DefaultPlatform() {
+		os, arch, err := parsePlatform(pf)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse platform %s", pf)
+		}
 		ctx.ArchitectureChoice = arch
+		ctx.OSChoice = os
```

Comment:
We shouldn't use `(*types.SystemContext).{Architecture,OS}Choice` for _creating_ an image. Those fields should only be used for _pulling_ an image (see https://github.com/containers/image/blob/master/types/types.go#L510).

Reply:
This illustrates what I was referring to before. `--override-{arch,os}` is used for pulling. When creating an image, we should use different options to prevent mixing things or blurring the lines.

Reply:
Hm, we still have to set it in the systemcontext, otherwise we would get the error on commit:
```
> ./buildah bud --pull-never --arch=arm -f tests/bud/platform/Dockerfile -t test
STEP 1: FROM scratch
STEP 2: COMMIT test
error committing base container: error copying layers and metadata for container "23ffa04e08322148180e7e45a39804ee9c2663758e0cfcd3a9bccaa2ca4f0333": Image operating system mismatch: image uses "arm", expecting "linux"
ERRO exit status 1
```

This issue is caused by the system context verification here:
https://github.com/containers/image/blob/b0893d83e0e450317ef781e9d6a621182d1a6346/copy/copy.go#L715-L721

Reply:
` image uses "arm", expecting "linux"` is quite wrong in many ways.

I assume that we need to set the values when committing the container, which internally uses image/copy.

Reply:
Yes, the first issue is caused by the os/arch flip error.